pkg/image/apis/image/validation/whitelist/whitelister.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (rw *registryWhitelister) AdmitDockerImageReference(ref *imageapi.DockerImageReference, transport WhitelistTransport) error {
+	const showMax = 5
+	const fullWhitelistLogLevel = 5
+
+	if rw.pullSpecs.Len() > 0 {
+		if rw.pullSpecs.Has(ref.Exact()) || rw.pullSpecs.Has(ref.DockerClientDefaults().Exact()) || rw.pullSpecs.Has(ref.DaemonMinimal().Exact()) {
+			return nil
+		}
+	}
+
+	if rw.registryHostRetriever != nil {
+		if localRegistry, ok := rw.registryHostRetriever.InternalRegistryHostname(); ok {
+			rw.WhitelistRegistry(localRegistry, WhitelistTransportSecure)
+		}
+	}
+
+	var (
+		host, port string
+		err        error
+	)
+	switch transport {
+	case WhitelistTransportAny:
+		host, port, err = net.SplitHostPort(ref.Registry)
+		if err != nil || len(port) == 0 {
+			host = ref.Registry
+			port = ""
+		}
+		if len(host) == 0 {
+			host, _ = ref.RegistryHostPort(false)
+		}
+	case WhitelistTransportInsecure:
+		host, port = ref.RegistryHostPort(true)
+	default:
+		host, port = ref.RegistryHostPort(false)
+	}
+
+	matchHost := func(h string) bool {
+		for _, hp := range rw.whitelist {
+			if stringsutil.IsWildcardMatch(h, hp.host) {
+				if len(port) == 0 {
+					switch hp.port {
+					case "80", "443", "*":
+						return true
+					default:
+						continue
+					}
+				}
+				if stringsutil.IsWildcardMatch(port, hp.port) {
+					return true
+				}
+			}
+		}
+		return false
+	}
+
+	switch host {
+	case imageapi.DockerDefaultV1Registry, imageapi.DockerDefaultV2Registry:
+		// try to match plain docker.io first to satisfy `docker.io:*` wildcard
+		if matchHost(imageapi.DockerDefaultRegistry) {
+			return nil
+		}
+		fallthrough
+	default:
+		if matchHost(host) {
+			return nil
+		}
+	}
+
+	hostname := ref.Registry
+	if len(ref.Registry) == 0 {
+		if len(port) > 0 {
+			hostname = net.JoinHostPort(host, port)
+		} else {
+			hostname = host
+		}
+	}
+	whitelist := []string{}
+	full := []string{}
+	for i := 0; i < len(rw.whitelist); i++ {
+		if i < showMax-1 || len(rw.whitelist)-showMax == 0 {
+			whitelist = append(whitelist, fmt.Sprintf("%q", net.JoinHostPort(rw.whitelist[i].host, rw.whitelist[i].port)))
+		} else if i < showMax {
+			whitelist = append(whitelist, fmt.Sprintf("and %d more ...", len(rw.whitelist)-showMax+1))
+		} else if !glog.V(fullWhitelistLogLevel) {
+			break
+		}
+		full = append(full, fmt.Sprintf("%q", net.JoinHostPort(rw.whitelist[i].host, rw.whitelist[i].port)))
+	}
+	glog.V(fullWhitelistLogLevel).Infof("registry %q not allowed by whitelist { %s }", hostname, strings.Join(full, ", "))
+	return fmt.Errorf("registry %q not allowed by whitelist { %s }", hostname, strings.Join(whitelist, ", "))
+}
+
+func (rw *registryWhitelister) WhitelistRegistry(hostPortGlob string, transport WhitelistTransport) error {
+	hp := allowedHostPortGlobs{}
+
+	parts := strings.SplitN(hostPortGlob, ":", 3)
+	switch len(parts) {
+	case 1:
+		hp.host = parts[0]
+		switch transport {
+		case WhitelistTransportAny:
+			hp.port = "*"
```

Comment:
I've just realized this is way too benevolent. `any` should mean either *secure* (`:443`) or *insecure* (`:80`) not *"any transport protocol"*. I'll fix this.